### PR TITLE
Option to Raise FILESYSTEM and WEBSCREENSHOT with base64 blobs

### PR DESCRIPTION
**Pull request by nicpenning**
Draft code things

### CODE REVIEW

**Comment by nicpenning** [comment]
FYI - Needs work and is not currently functioning or well thought out.

**Comment by TheTechromancer** [comment]
It's a good start. Tomorrow I'll see if I can pitch in and get it working.

**Comment by TheTechromancer** [comment]
@nicpenning are you able to allow edits from maintainers?

image

**Comment by nicpenning** [comment]
image

Looks like I already had that set.

image

**Comment by TheTechromancer** [comment]
Okay, I made a few changes:

- Added global `file_blobs` and `folder_blobs` options
- Wrote tests
- Added feature for both `WEBSCREENSHOT` and `FILESYSTEM`

**Comment by nicpenning** [comment]
Phenomenal!

**Comment by nicpenning** [comment]
Please let me know how I can test with custom/flag options, and I can run do some trial runs to see how it shakes out.

**Comment by TheTechromancer** [comment]
> Please let me know how I can test

@nicpenning you can test like this:

```bash
bbot -t evilcorp.com -c file_blobs=true folder_blobs=true
```

This will automatically attach binary blobs to any `WEBSCREENSHOT` and `FILESYSTEM` events. Do be aware that if the `FILESYSTEM` event is a folder (i.e. a git repo), the blob contains a compressed tar.gz archive.

File: bbot/modules/git_clone.py [hunk 1]
```diff
@@ -39,9 +39,7 @@ async def handle_event(self, event):
         repo_path = await self.clone_git_repository(repo_url)
         if repo_path:
             self.verbose(f"Cloned {repo_url} to {repo_path}")
-            codebase_event = self.make_event(
-                {"path": str(repo_path)}, "FILESYSTEM", tags=["git", "folder"], source=event
-            )
+            codebase_event = self.make_event({"path": str(repo_path)}, "FILESYSTEM", tags=["git"], source=event)
```
Comment (domwhewell-sage) [inline bbot/modules/git_clone.py:42]: The removal of the folder tag here will stop the unstructured module from crawling the folder

Comment (TheTechromancer) [reply bbot/modules/git_clone.py:42]: The tag will still be there, only now it's added automatically during the creation of the event.

**Comment by TheTechromancer** [comment]
@nicpenning have you got a chance to test?

**Comment by nicpenning** [comment]
Not yet, but I will test as soon as I can! My initial test didn't work but I might not have had my config options or code updated properly.

**Comment by nicpenning** [comment]
So far works great! Only downside is that the blob event that gets emitted to the console takes up a lot of lines (like 600K and in some cases I imagine it could be millions). It is bearable, but makes it more difficult to watch the scans progression using the default output.

Going to test the strict options to ensure I only get blobs for the appropriate scope.

**Comment by nicpenning** [comment]
So does FILESYSTEM == filedownload?

I get all of the screenshots as PNGs in the output.ndjson but never the filedownload but shouldn't I need to set that configuration for the filedownload to use the blob feature? It doesn't seem the pdf, docx, etc.. that are in the filedownload directory are in the event at all (not just missing the blob).

Running this command `bbot -t evilcorp.com
-c file_blobs=true folder_blobs=true output_modules.http.siem_friendly=true modules.filedownload.base_64_encoded_file=true -m filedownload httpx gowitness`


```
{"type": "WEBSCREENSHOT", "id": "WEBSCREENSHOT:4ec4b0a8e3ce89c7a85d9bb9b81d91dd652cccea", "data": {"filename": "/home/yams/.bbot/scans/heightened_jeremy/gowitness/screenshots/https-evilcorp.com.png", "url": "https://evilcorp.com/inside-evilcorp", "blob": "iVBORw0KG..."}
```

TL:DR - Screenshots seem to work, but the FILESYSTEM / filedownload module does not emit an event (so blob definitely does not exist).

**Comment by nicpenning** [comment]
Here is a real world example: bbot -t "https://eicar.org/download-anti-malware-testfile/" -c file_blobs=true folder_blobs=true output_modules.http.siem_friendly=true modules.filedownload.base_64_encoded_file=true -m filedownload httpx gowitness

Files downloaded, but none are in the events (output.ndjson).
image

Otherwise, everything else is working as expected.

**Comment by TheTechromancer** [comment]
Hmm I think this branch needs merged with the latest dev, so it includes this PR:

**Comment by TheTechromancer** [comment]
I'm going to go ahead and merge this. We can make tweaks in a followup PR if needed.

**Review by TheTechromancer — APPROVED** [review]

**Comment by nicpenning** [comment]
Sounds great

**Review by liquidsec — APPROVED** [review]
